framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,839 @@
+        for (QualifierKind qualifierKind : nameToQualifierKind.values()) {
+            boolean isPoly = qualifierKind.isPoly;
+            boolean hasSubtypeOfAnno = directSuperMap.containsKey(qualifierKind);
+            if (isPoly && hasSubtypeOfAnno) {
+                // This is currently not supported. At some point we might add
+                // polymorphic qualifiers with upper and lower bounds.
+                throw new UserError(
+                        "AnnotatedTypeFactory: "
+                                + qualifierKind
+                                + " is polymorphic and specifies super qualifiers.%n"
+                                + "Remove the @PolymorphicQualifier or @SubtypeOf annotation from it.");
+            } else if (!isPoly && !hasSubtypeOfAnno) {
+                throw new UserError(
+                        "AnnotatedTypeFactory: %s does not specify its super qualifiers.%n"
+                                + "Add an @SubtypeOf or @PolymorphicQualifier annotation to it,%n"
+                                + "or if it is an alias, exclude it from `createSupportedTypeQualifiers()`.%n",
+                        qualifierKind);
+            } else if (isPoly) {
+                if (qualifierKind.top == null) {
+                    throw new UserError(
+                            "PolymorphicQualifier, %s,  has to specify type hierarchy, if more than one exist; top types: [%s] ",
+                            qualifierKind, SystemUtil.join(", ", tops));
+                } else if (!tops.contains(qualifierKind.top)) {
+                    throw new UserError(
+                            "Polymorphic qualifier %s has invalid top %s. Top qualifiers: %s",
+                            qualifierKind, qualifierKind.top, SystemUtil.join(", ", tops));
+                }
+            }
+        }
+
+        if (bottoms.size() != tops.size()) {
+            throw new UserError(
+                    "Number of tops not equal to number of bottoms: Tops: [%s] Bottoms: [%s]",
+                    SystemUtil.join(", ", tops), SystemUtil.join(", ", bottoms));
+        }
+    }
+
+    /**
+     * Creates all QualifierKind objects for the given qualifier classes and adds them to
+     * qualifierClassMap. This method does not initialize all fields in the {@link QualifierKind};
+     * that is done by {@link #initializeQualifierKindFields(Map)}.
+     *
+     * @param qualifierClasses classes of annotations that are type qualifiers
+     * @return a mapping from the canonical name of an annotation class to {@link QualifierKind}
+     */
+    protected NavigableMap<@Interned String, QualifierKind> createQualifierKinds(
+            Collection<Class<? extends Annotation>> qualifierClasses) {
+        TreeMap<@Interned String, QualifierKind> nameToQualifierKind = new TreeMap<>();
+        for (Class<? extends Annotation> clazz : qualifierClasses) {
+            @SuppressWarnings("interning") // uniqueness is tested immediately below
```

Comment:
Why is this required?  Since the QualifierKind class is annotated as `@Interned`, I expect its constructor to return an `@Interned` object.  Am I incorrect about that?

Reply:
The type of the constructor result is `@UnknownInterned`.  See [the manual](https://checkerframework.org/manual/#interning-interned-classes).